mfr/core/extension.py

```diff
@@ -15,38 +29,90 @@ def __init__(self, ext, source_file_path, output_file_path, format):
         :param format: the format of the exported file (e.g. 1200*1200.jpg)
         """
 
-        self.ext = ext
-        self.source_file_path = source_file_path
-        self.output_file_path = output_file_path
+        self.metadata = metadata
+        self.ext = metadata.ext
+        self.input_stream = input_stream
         self.format = format
+        self.source_file_id = uuid.uuid4()
+        self.cache_file_id = '{}.{}'.format(self.metadata.unique_key, self.format)
+        self.local_cache_provider = waterbutler.core.utils.make_provider(
+            'filesystem', {}, {}, settings.LOCAL_CACHE_PROVIDER_SETTINGS
+        )
+        self.exporter_name = utils.get_exporter_name(self.metadata.ext)
+        if self.exporter_name:
+            cache_file_path_str = '/export/{}.{}'.format(self.cache_file_id, self.exporter_name)
+        else:
+            cache_file_path_str = '/export/{}'.format(self.cache_file_id)
+
         self.exporter_metrics = MetricsRecord('exporter')
         if self._get_module_name():
             self.metrics = self.exporter_metrics.new_subrecord(self._get_module_name())
 
+    async def __call__(self):
+
+        self.source_wb_path = await self.local_cache_provider.validate_path(
+            '/export/{}'.format(self.source_file_id)
+        )
+        self.source_file_path = self.source_wb_path.full_path
+        self.output_file_id = '{}.{}'.format(self.source_wb_path.name, self.format)
+        self.output_wb_path = await self.local_cache_provider.validate_path(
+            '/export/{}'.format(self.output_file_id)
+        )
+        self.output_file_path = self.output_wb_path.full_path
         self.exporter_metrics.merge({
             'class': self._get_module_name(),
             'format': self.format,
-            'source_path': str(self.source_file_path),
-            'output_path': str(self.output_file_path),
+            'source_path': str(self.source_wb_path),
+            'output_path': str(self.output_wb_path),
             # 'error': 'error_t',
             # 'elapsed': 'elpased_t',
         })
 
+        # Put the file into the local cahe provider (local fs)
+        await self.local_cache_provider.upload(
+            self.input_stream,
+            self.source_wb_path
+        )
+
+        # Note where the output file is for diagnostics
+        self.exporter_metrics.merge({
+            'output_file': {
+                'id': self.output_file_id,
+                'path': str(self.output_wb_path),
+                'provider': self.local_cache_provider.NAME,
+            }
+        })
+
+        # Execute the extension's export method asynchronously and wait for it
+        # to finish
+        loop = asyncio.get_event_loop()
+        await loop.run_in_executor(None, self.export)
+
+        # Return a stream of the converted file
+        return await self.write_to_stream()
+
+    def __del__(self):
+        self.output_fp.close()
```

Comment:
This I need a better solution for - this only works because the exporter goes out of scope when the request finishes. It works and will be reliable, but I'd like something a little more pretty.

Reply:
I think I mentioned this somewhere above, is this really necessary? Why context manager doesn't work in your case?